tests/testlib/s2n_testlib.h

```diff
@@ -245,3 +245,40 @@ extern const s2n_parsed_extension EMPTY_PARSED_EXTENSIONS[S2N_PARSED_EXTENSIONS_
 int s2n_kem_recv_public_key_fuzz_test(const uint8_t *buf, size_t len, struct s2n_kem_params *kem_params);
 int s2n_kem_recv_ciphertext_fuzz_test(const uint8_t *buf, size_t len, struct s2n_kem_params *kem_params);
 int s2n_kem_recv_ciphertext_fuzz_test_init(const char *kat_file_path, struct s2n_kem_params *kem_params);
+
+/* kTLS */
+#define S2N_TEST_KTLS_MOCK_HEADER_SIZE        3
+#define S2N_TEST_KTLS_MOCK_HEADER_LENGTH_SIZE 2
+
+/* The record_type is communicated via ancillary data when using kTLS. For this
+ * reason s2n must use `send/recvmsg` syscalls rather than `send/read`. To mimic
+ * the send/recvmsg calls more accurately, we mock the socket via two separate
+ * buffers: data_buffer and ancillary_buffer.
+ *
+ * The mock implementation, uses 3 bytes with a tag + len format to represent
+ * each record. The first byte(tag) is the record_type and the next two represent
+ * the length of the record. Length is represented as a u16 to capture the max
+ * possible TLS record length.
+ *
+ * ancillary_buffer memory layout:
+ *    [      u8      |            u16            ]
+ *      record_type             length
+ */
+struct s2n_test_ktls_io_stuffer {
+    struct s2n_stuffer ancillary_buffer;
+    struct s2n_stuffer data_buffer;
+    size_t invoked_count;
+};
+struct s2n_test_ktls_io_pair {
+    struct s2n_test_ktls_io_stuffer client_in;
+    struct s2n_test_ktls_io_stuffer server_in;
+};
+ssize_t s2n_test_ktls_sendmsg_stuffer_io(struct s2n_connection *conn, struct msghdr *msg, uint8_t record_type);
+ssize_t s2n_test_ktls_recvmsg_stuffer_io(struct s2n_connection *conn, struct msghdr *msg, uint8_t *record_type);
+
+S2N_RESULT s2n_test_init_ktls_stuffer_io(struct s2n_connection *server, struct s2n_connection *client,
+        struct s2n_test_ktls_io_pair *io_pair);
```

Comment:
Maybe put this stuff in a separate header file? There's no reason why any other test would need to use these functions.

Reply:
Originally had it only in the test file but I actually plan to reuse all these in the next few PRs to setup and validate: https://github.com/aws/s2n-tls/pull/4119